open_schools_platform/organization_management/circles/services.py

```diff
@@ -50,42 +52,79 @@ def create_circle(name: str, organization: Optional[Organization], description:
+        (Query.Status.SENT, frozenset({Family, Circle})): (
+            Query.Status.DECLINED, Query.Status.CANCELED, Query.Status.ACCEPTED),
+    }
+    change_query: Dict[Type, Callable[[Query, str], Query]] = {
+        Family: query_to_family,
+        TeacherProfile: query_to_teacher_profile
+    }
 
     @staticmethod
     def query_handler(query: Query, new_status: str, user: User):
         BaseQueryHandler.query_handler_checks(CircleQueryHandler, query, new_status, user)
 
-        from open_schools_platform.parent_management.families.models import Family
+        change_query_function = CircleQueryHandler.change_query.get(type(query.recipient), query_wrong_recipient)
+        user_role = CircleQueryHandler._parse_changer_type(user, query)
 
-        if type(query.recipient) is not Family:
-            raise ValidationError(detail="The recipient must be a Family if the sender is a circle")
-
-        circle_access = user.has_perm('circles.circle_access', query.sender)
-        family_access = user.has_perm('families.family_access', query.recipient)
-
-        if not family_access:
-            if new_status != Query.Status.CANCELED:
-                raise NotAcceptable("Circle can only set canceled status")
-        elif not circle_access:
-            if query.status != Query.Status.SENT:
-                raise NotAcceptable("Сan no longer change the query")
-            if new_status == Query.Status.CANCELED:
-                raise NotAcceptable("User cannot cancel query, he can only decline or accept it")
-
-        query_update(query=query, data={"status": new_status})
-        if query.status == Query.Status.ACCEPTED:
-            if query.body is None:
-                raise MethodNotAllowed("put", detail="Query is corrupted")
-            query.body.circle = query.sender
-            query.body.student_profile = query.additional
-            query.body.save()
-
-        return query
+        if new_status in CircleQueryHandler.available_statuses[(query.status, user_role)]:
+            return change_query_function(query, new_status)
+        else:
+            raise NotAcceptable(f"Status change [{query.status} => {new_status}] not allowed for {user_role}")
 
     setattr(Circle, "query_handler", query_handler)
 
+    @staticmethod
+    def _parse_changer_type(user: User, query: Query) -> frozenset:
+        circle_access = user.has_perm('circles.circle_access', query.sender)
+        family_access = type(query.recipient) == Family and user.has_perm('families.family_access', query.recipient)
+        teacher_profile_access = type(query.recipient) == TeacherProfile and user.has_perm(
+            "teachers.teacher_profile_access", query.recipient)
+        changer_type = set()
+        if circle_access:
+            changer_type.add(Circle)
+        if family_access:
+            changer_type.add(Family)
+        if teacher_profile_access:
+            changer_type.add(TeacherProfile)
```

Comment:
it is ok but make it so if follow after access
it is need for new developers, they can forget to write if
you can come up better solution to not write ifs for each access